# 2. Running Time & Big O Notation

## Running Time [1/2]
- Generally you want to choose the most efficient algorithm; wheter you're trying to optimize for time or space.
- Back to [Binary Search](http://localhost:8888/notebooks/01_searching_algorithm.ipynb?). How much do you save by using it?
- Well, the first approach was to check each number, one by one (**Simple Search**). If this is a list of 100 numbers, it takes up to 100 guesses. If it's a list of 4 billion numbers, it takes up to 4 billion guesses.
- So the maximum number of guesses is the same as the size of the list. This is called **Linear Time**.
- **Binary Search** is different. If the list is 100 items long, it takes at most 7 guesses. If the list is 4 billion items, it takes up to 32 guessess.
- **Binary Search** runs **logarithmic time** (or log time, as the natives call it).

## Running Time [2/2]
Run times for search algorithms

![Logo](https://drek4537l1klr.cloudfront.net/bhargava2/Figures/image_1-16.png)

## Big O Notation
Big O Notation is a special notation that tells you how fast an algorithm is when the size of the input grows.

![Logo](https://drek4537l1klr.cloudfront.net/bhargava2/Figures/image_1-15.png)

## Algorithm running times grow at different rates [1/6]
- Bob is writing an algorithm for NASA.
- His algorithm will kick in when a rocket is about to land on the Moon, and it will help calculate where to land.
- Bob has only 10 seconds to figure out where to land

![Logo](https://drek4537l1klr.cloudfront.net/bhargava2/Figures/1.png)

## Algorithm running times grow at different rates [2/6]
- Bob is trying to decide between **Simple Search** and **Binary Search**.
- The algorithm needs to be fast and correct.
- On one hand, **Binary Search** is faster.
- On the other hand, **Simple Search** is easier to write, and there is a less chance of bugs being introduced. And Bob doesn't want bugs in the code to land a rocket!

## Algorithm running times grow at different rates [3/6]
- Let's assume it takse 1 milisecond to check one element.
- To be extra careful, Bob decides to time both algorithms with a list of 100 elements.
- With **Simple Search**, Bob has to check 100 elements, so the search takes 100ms to run.
- On the other hand, he only has to check 7 elements with **Binary Search** ($log_{2} 100$ is roughly 7), so that search takes 7ms to run.

![Logo](https://drek4537l1klr.cloudfront.net/bhargava2/Figures/image_1-18.png)

## Algorithm running times grow at different rates [4/6]
- But realisticly, the list wil have more like a billion elements.
- Bob runs **Binary Search** with 1 billion elements, and it takes 30ms ($log_{2} 1,000,0000,000$ is roughly 30).
- **Binary Search** is about 15ms times faster than **Simple Search**, because **Simple Search** took 100ms with 100 elements, and **Binary Search** took 7ms.
- So **Simple Search** will take 30 x 15 = 450ms, right? Way under my threshold of 10 seconds.
- Bob decides to go with **Simple Search**.
- Is that the right choice?

## Algorithm running times grow at different rates [5/6]
- No! Turns out, Bob is wrong.
- The run time for **Simple Search** with 1 billion elements will be 1 billion ms, which is 11 days!
- The problem is, the run times for **Binary Search** and **Simple Search** don't grow at the same rate.

![Logo](https://drek4537l1klr.cloudfront.net/bhargava2/Figures/image_1-19.png)

## Algorithm running times grow at different rates [6/6]
- As the number of items increases, **Binary Search** takes a litte more time to run, but **Simple Search** takes a lot more time to run.
- So as the list of numbers gets bigger, **Binary Search** suddenly becomes more faster than **Simple Search**.
- That's why it's not enough to know how long an algorithm takes to run, you need to know how the running time increases as the list size increases.
- That's where Big O Notation comes in!

## Big O Notation explained [1/2]
- Big O Notation tells you how fast an algorithm is.
- For example, suppose you have a list size of $n$.
- **Simple Search** needs to check each element, so it will take $n$ operations. The run time in Big O Notation is $O(n)$,
- Big O doesn't tell you the speed in seconds. Big O Notation lets you compare the number of operations. It tells you how fast the algorithm grows.

## Big O Notation explained [2/2]
- **Binary Search** needs $log n$ operations to check a list of size $n$.
- What's the running time in Big O Notation? It's $O(log n)$.
- In general, Big O Notation written as follows:

![Logo](https://drek4537l1klr.cloudfront.net/bhargava2/Figures/image_1-21.png)

- This tells you the number of operations an algorithm will make.
- It's called Big O Notation, because you put a "big O" in front of the number operations.

## Code examples

In [2]:
# Simple Search algorithm code
def simple_search(list,item):
    for num in range(len(list)):
        if list[num] == item:
            return num
    return None

In [4]:
%timeit first_result = simple_search(list(range(1024 + 1)),1024)

69.7 μs ± 5.47 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


## Output explainations [1/2]
- $69.7 \, \mu\text{s}$: The mean (average) time taken to execute the block once. The symbol "$\mu\text{s}$"" stands for microseconds.
- $\pm 5.47 \, \text{ns}$: This is the standard deviation (a measure of variability) of the timings. It indicates how much the individual measurements deviate from the mean. The symbol "$\text{ns}$" stands for nanoseconds.
- per loop: This specifies that the timings are measured per iteration of the loop.
- (mean + std. dev. of 7 runs, 10,000 loops each): This indicates that the banchmark was run 7 times, and each run consisted of 10,000 iterations of the loop. The mean and the standard deviation are calculated based on these 7 runs.

> **See Also**: For more in-depth about **%timeit** magic command and any kind of magic commands, [you can click here](https://jakevdp.github.io/PythonDataScienceHandbook/01.03-magic-commands.html).


In [6]:
def binary_search(list,item):
    low = 0
    high = len(list) - 1

    while low <= high:
        mid = (low + high) // 2
        guess = list[mid]
        if guess == item:
            return mid
        if guess > item:
            mid = mid - 1
        else:
            mid = mid + 1
    return None

In [8]:
%timeit second_result = simple_search(list(range(1024 + 1)),1024)

67.7 μs ± 2.72 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


## Output explainations [2/2]
- $67.7 \, \mu\text{s}$: On average, each execution of the **binary_search** function took 15.6 microseconds.
- $\pm\ 2.72 \, \text{ns}$: The standard deviation of the execution time is 69.6 seconds. Measures the varibility in the time taken per loop.
- 7 runs: The timing result is based on 7 seperate runs of the **%timeit** magic commands.
- 100,000 loops: Withing each run, the code was executed 100,000 times to get a reliable average timing.

## Visualzing different Big O run times (another simple cases study)
- Here's a practical example that you can follow at home with a few pieces of paper and a pencil.
- Suppose you have to draw a grid of 16 boxes.

![Logo](https://drek4537l1klr.cloudfront.net/bhargava2/Figures/image_1-22.png)

- What's a good algorithm to draw this grid?

## Algorithm 1
- One way to do it is to draw 16 boxes, one at a time.
- Remember, Big O Notation counts the number of operations.
- In this example, drawing one box is one operation. You have to draw 16 boxes.
- How many operations it will take, drawing one box at a time?

![Logo](https://drek4537l1klr.cloudfront.net/bhargava2/Figures/image_1-23.png)

- 16 steps to draw 16 boxes. What's the running time for this algorithm?

## Algorithm 2 [1/3]
- Try this algorithm instead! Fold the paper.

![Logo](https://drek4537l1klr.cloudfront.net/bhargava2/Figures/image_1-24.png)

- In this example, folding the paper once is an operation.
- You just made two boxes with that operation!

## Algorithm 2 [2/3]
- Folder the paper again, again, and again.

![Logo](https://drek4537l1klr.cloudfront.net/bhargava2/Figures/image_1-25.png)

- Unfold it after four fold, and you'll have a beautiful grid!

## Algorithm 2 [3/3]
![Logo](https://drek4537l1klr.cloudfront.net/bhargava2/Figures/image_1-26.png)

- Every fold doubles the number of boxes. You made 16 boxes with 4 operations!
- You can "draw" twice as many boxes with every fold, so you can draw 16 boxes in 4 steps.
- What's the running time for this algorithm?
- Come up with running times for both algorithms before moving on.
- Answers: Algorithm 1 takes $O(n)$ time, and algorithm 2 takes $O(log_{2} n)$ time.

## Big O establishes a worst-case run time [1/2]
- Suppose you're using **Simple Search** to look for a person in a phone book.
- You know that **Simple Search** takes $O(n)$ time to run, which means in the worst case, you'll have to look through every single entry in the phone book.
- In this case, you're looking for Adit. This guy is the first entry in the phone book. So you didn't have to look at every entry, you found it on the first try.
- Did this algorithm take $O(n)$ time? Or did it take $O(1)$ time, because you found the person on the first try?

## Big O establishes a worst-case run time [2/2]
- **Simple Search** still takes $O(n)$ time.
- In this case, you found what you were looking for instantly. That's the **best-case** scenario.
- But **Big O Notation** is all about the **worst-case** scenario.
- So you can say that, in the worst case, you'll need to look at every entry in the phone book once. That's $O(n)$ time.
- It's a reassurance, you know that **Simple Search** will never be slower than $O(n)$ time.

## Some common Big O run times [1/4]
Here are five Big O run times that you'll encounter a lot, sorter from fastest to slowest:
- $O(log n)$: Also known as **logarithmic time** complexity. This means the runtime grows logarithmically with the input size. Example: **Binary Search**.
- $O(n)$: Also known as **linear time** complexity. This means the runtime grows linearly with the input size. Example: **Simple Search**.
- $O(n * log n)$, also known as **linearithmic time** complexity. This means the runtime grows linearithmically with the input size. Example: A fast sorting algorithm, like quicksort (coming up in chapter 4).
- $O(n^2)$, also known as **quadratic time** complexity. This means the runtime grows quadratically with the input size Example: A slow sorting algorithm, like selection sort (coming up in chapter 2).
- $O(n!)$, also known as **factorial time** complexity. This means the runtime  grows factorially with the input size. Example: A really slow sorting algorithm, like the traveling sales person (coming up next!).

## Some common Big O run times [2/4]
- Suppose you're drawing a grid of 16 boxes again, and you can choose from 5 different algorithms to do so.
- You can do 10 operations per second.
- If you use the first algorithm, it will take you $O(log n)$ time to draw the grid.
- With $O(log n)$ time, it will take you 4 operations to draw a grid of 16 boxes ($log_{2} 16$ is 4). So it will take 0.4 seconds to draw the grid.
- What if you have to draw 1,024 boxes? It will take you $log_{2} 1,024 = 10$ operations, or 1 second to draw a grid of 1,024 boxes.
- These numbers are using the first algorithm.

## Some common Big O run times [3/4]
- The second algorithm is slower: it takes $O(n)$ time.
- It will take 16 operations to draw 16 boxes, and it will take 1,024 operations to draw 1,024 boxes.
- How much time is that in seconds?

## Some common Big O run times [4/4]
- Here's how long it would take to draw a grid for the rest algorithms, from fastest to slowest:
![Logo](https://drek4537l1klr.cloudfront.net/bhargava/Figures/016fig01_alt.jpg)
- There are other run times too, but these are the five mossot common run times.
- This is a simplification. In reality you can't convert from a Big O run time to a numbern of operations neatly, but this is good enough for now.

## Big O Notation (recap)
For now, the main takeaways are as follows:
- Algorithm speed isn't measured in seconds, but in growth of the number of operations.
- Instead, we talk about how quickly the run time of an algorithm increases as the size of the input increases.
- Run time of an algorithms are expressed in Big O Notation.
- In general, $O(log n)$ is faster than $O(n)$, but it gets a lot faster as the list of the item grows.